src/creatures/combat/combat.cpp

```diff
@@ -251,190 +251,133 @@ ReturnValue Combat::canTargetCreature(const std::shared_ptr<Player> &player, con
 	return canDoCombat(player, target, true);
 }
 
-ReturnValue Combat::canDoCombat(const std::shared_ptr<Creature> &caster, const std::shared_ptr<Tile> &tile, bool aggressive) {
-	if (!aggressive) {
-		return RETURNVALUE_NOERROR;
+bool Combat::isInPvpZone(const std::shared_ptr<Creature> &attacker, const std::shared_ptr<Creature> &target) {
+	return attacker->getZoneType() == ZONE_PVP && target->getZoneType() == ZONE_PVP;
+}
+
+bool Combat::isProtected(const std::shared_ptr<Player> &attacker, const std::shared_ptr<Player> &target) {
+	uint32_t protectionLevel = g_configManager().getNumber(PROTECTION_LEVEL);
+	if (target->getLevel() < protectionLevel || attacker->getLevel() < protectionLevel) {
+		return true;
 	}
 
-	if (tile->hasProperty(CONST_PROP_BLOCKPROJECTILE)) {
-		bool canThrow = false;
+	if ((!attacker->getVocation()->canCombat() || !target->getVocation()->canCombat()) && (attacker->getVocationId() == VOCATION_NONE || target->getVocationId() == VOCATION_NONE)) {
+		return true;
+	}
 
-		if (const auto fieldList = tile->getItemList()) {
-			for (const auto &findfield : *fieldList) {
-				if (findfield && (findfield->getID() == ITEM_MAGICWALL || findfield->getID() == ITEM_MAGICWALL_SAFE)) {
-					canThrow = true;
-					break;
-				}
-			}
-		}
+	if (attacker->getSkull() == SKULL_BLACK && attacker->getSkullClient(target) == SKULL_NONE) {
+		return true;
+	}
 
-		if (!canThrow) {
-			return RETURNVALUE_CANNOTTHROW;
-		}
+	return false;
+}
+
+ReturnValue Combat::canDoCombat(const std::shared_ptr<Creature> &caster, const std::shared_ptr<Tile> &tile, bool aggressive) {
+	if (!tile || !aggressive) {
+		return RETURNVALUE_NOERROR;
 	}
 
-	if (aggressive && tile->hasFlag(TILESTATE_PROTECTIONZONE)) {
+	if (tile->hasFlag(TILESTATE_PROTECTIONZONE)) {
 		return RETURNVALUE_ACTIONNOTPERMITTEDINPROTECTIONZONE;
 	}
 
+	if (tile->hasProperty(CONST_PROP_BLOCKPROJECTILE)) {
+		const auto fieldList = tile->getItemList();
+		if (!fieldList || std::none_of(fieldList->begin(), fieldList->end(), [](const auto &item) {
+				return item && (item->getID() == ITEM_MAGICWALL || item->getID() == ITEM_MAGICWALL_SAFE);
+			})) {
+			return RETURNVALUE_CANNOTTHROW;
+		}
+	}
+
 	if (tile->getTeleportItem()) {
 		return RETURNVALUE_CANNOTTHROW;
 	}
 
 	if (caster) {
-		const Position &casterPosition = caster->getPosition();
-		const Position &tilePosition = tile->getPosition();
+		const auto &casterPosition = caster->getPosition();
+		const auto &tilePosition = tile->getPosition();
 		if (casterPosition.z < tilePosition.z) {
 			return RETURNVALUE_FIRSTGODOWNSTAIRS;
 		} else if (casterPosition.z > tilePosition.z) {
 			return RETURNVALUE_FIRSTGOUPSTAIRS;
 		}
 
-		if (const auto &player = caster->getPlayer()) {
-			if (player->hasFlag(PlayerFlags_t::IgnoreProtectionZone)) {
-				return RETURNVALUE_NOERROR;
-			}
+		const auto player = caster->getPlayer();
```

Comment:
Declare as const ref (&), to avoid increment reference counter (same the others variables)
```suggestion
		const auto &player = caster->getPlayer();
```

Reply:
Ins't?
const auto& player = caster->getPlayer();

Reply:
Now it's time to make lunch. Later, I can come back and check if everything is in order.
I updated it as you mentioned, and I believe it's good.
Thanks for the tip!

#3357